### PR TITLE
 test/e2e: Introduce a e2e framework for testing - part 2

### PR DESCRIPTION
**Pull request by wainersm**
This PR contains some fixes and features missed in the https://github.com/confidential-containers/cloud-api-adaptor/pull/386 that introduced the e2e framework.

### CODE REVIEW

File: test/e2e/README.md [hunk 1]
```diff
@@ -26,6 +26,10 @@ The `TEST_E2E_PODVM_IMAGE` is an optional variable which specifies the path to t
 $ TEST_E2E_PROVISION=yes TEST_E2E_PODVM_IMAGE="path/to/podvm-base.qcow2" CLOUD_PROVIDER=libvirt make test-e2e
 ```
 
+By default it is given 20 minutes for the entire e2e execution to complete, otherwise the process is preempted. If you need to extend that timeout then export the `TEST_E2E_TIMEOUT` variable. For example, `TEST_E2E_TIMEOUT=30m` set the timeout to 30 minutes. See `-timeout` in [go test flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) for the values accepted.
+
+To leave the cluster untouched by the execution finish you should export `TEST_E2E_TEARDOWN=no`, otherwise the framework will attempt to wipe out any resources created. For example, if `TEST_E2E_PROVISION=yes` then it will delete both the cluster and VPC.
+
```
Comment (bpradipt) [inline test/e2e/README.md:32]: nit: TEST_E2E_PROVISION=yes => TEST_E2E_TEARDOWN=yes

**Comment by bpradipt** [comment]
@wainersm the name `PeerPods` for the struct handling cloud-api-adaptor (CAA) properties is confusing and overload. The PeerPod name refers to the pod. Once we have CRD to handle dangling resources the CRD's naming also most likely be PeerPod to reflect the pod.  But in e2e PeerPod refers to CAA.  

I think renaming the PeerPod struct to something like PeerPodInfra will help. 

@stevenhorsman @jensfr @kartikjoshi21 @snir911 any comments ?

File: test/e2e/README.md (hunk 1 above)
Comment (stevenhorsman) [reply test/e2e/README.md:32]: So I think Wainer's doc is actually correct, but maybe could do with some more clarification e.g.
> For example, if `TEST_E2E_PROVISION=yes` it used to create a VPC and cluster for testing and `TEST_E2E_TEARDOWN=no` not specified, then at the end of the test the the provisioned cluster and VPC, will both be deleted.

I'm not sure if that helps @bpradipt and if I've interpreted it correctly @wainersm?

**Comment by stevenhorsman** [comment]
> @wainersm the name `PeerPods` for the struct handling cloud-api-adaptor (CAA) properties is confusing and overload. The PeerPod name refers to the pod. Once we have CRD to handle dangling resources the CRD's naming also most likely be PeerPod to reflect the pod. But in e2e PeerPod refers to CAA.
> 
> I think renaming the PeerPod struct to something like PeerPodInfra will help.

Yes, I definitely understand where you are coming from here. `PeerPodInfra` is fine, but I wonder if `CloudApiAdaptor` would be as good/better - I'm trying to think of a good name that would encompass cert-manager, webhook, the CoCo operator controller and cloud-api-adaptor in a succinct way, which is tricky!

File: test/e2e/README.md (hunk 1 above)
Comment (bpradipt) [reply test/e2e/README.md:32]: Thanks @stevenhorsman, the example you gave makes sense.

**Comment by bpradipt** [comment]
> > @wainersm the name `PeerPods` for the struct handling cloud-api-adaptor (CAA) properties is confusing and overload. The PeerPod name refers to the pod. Once we have CRD to handle dangling resources the CRD's naming also most likely be PeerPod to reflect the pod. But in e2e PeerPod refers to CAA.
> > I think renaming the PeerPod struct to something like PeerPodInfra will help.
> 
> Yes, I definitely understand where you are coming from here. `PeerPodInfra` is fine, but I wonder if `CloudApiAdaptor` would be as good/better 

Yeah CloudAPIAdaptor is also fine and leaves no confusion :-). Some more thoughts below.

> - I'm trying to think of a good name that would encompass cert-manager, webhook, the CoCo operator controller and cloud-api-adaptor in a succinct way, which is tricky!

So today PeerPodConfig CRD encompasses CAA configuration, webhook and Pod VM extended resources. The corresponding controller is in the directory - `peer-pod-controller`. I think we should rename the directory to peerpod-config-controller. The code uses PeerPodConfig.

We can also change `PeerPodConfig` to `CloudAPIAdaptorConfig` based on the above suggestion. 
@jensfr ??

Now PeerPod is the name for the CRD (WIP) handling dangling resources and other aspects related to the actual K8s Pod.
`PeerPod` or `ShadowPod` (suggested earlier by @yoheiueda afaik) reflects the relationship with the K8s Pod.

**Review by stevenhorsman — COMMENTED** [review]
The code looks reasonable to me, but I've tried testing it on three different VMs and each time hit:
```
Deploy Peer Pods
Install the controller manager
Wait for the cc-operator-controller-manager deployment be available
Install CoCo and cloud-api-adaptor
Wait for the pod cc-operator-daemon-install-lv4sd be ready
Wait for the pod cloud-api-adaptor-daemonset-libvirt-ns2cv be ready
Check the kata runtimeclass exists
F0210 14:23:36.459155   78242 env.go:360] Setup failure: runtimeclasses.node.k8s.io "kata" not found
```
I think the issue is the runtime class is created slightly after the pods go into ready state, so we might need to add a wait for them to be created too?

e.g. here is a run showing this:
```
FAIL
make: *** [Makefile:66: test-e2e] Error 1
root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
No resources found
root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get pods -A
NAMESPACE                        NAME                                             READY   STATUS      RESTARTS      AGE
autorules                        autoruler-76cdf5f9f6-d2pfx                       1/1     Running     0             4m31s
confidential-containers-system   cc-operator-controller-manager-fbb5dcf9d-h27f7   2/2     Running     1 (67s ago)   112s
confidential-containers-system   cc-operator-daemon-install-mgkcb                 1/1     Running     0             97s
confidential-containers-system   cloud-api-adaptor-daemonset-libvirt-cr4nz        1/1     Running     0             97s
ingress-nginx                    ingress-nginx-admission-create-qgrt8             0/1     Completed   0             4m32s
ingress-nginx                    ingress-nginx-admission-patch-5rtg6              0/1     Completed   1             4m32s
ingress-nginx                    ingress-nginx-controller-64f79ddbcc-v8nj6        1/1     Running     0             4m32s
kube-flannel                     kube-flannel-ds-fwhmc                            1/1     Running     0             5m21s
kube-flannel                     kube-flannel-ds-n7xt7                            1/1     Running     0             2m42s
kube-system                      coredns-787d4945fb-882tr                         1/1     Running     0             5m21s
kube-system                      coredns-787d4945fb-lgrjq                         1/1     Running     0             5m21s
kube-system                      etcd-peer-pods-ctlplane-0                        1/1     Running     0             5m35s
kube-system                      kube-apiserver-peer-pods-ctlplane-0              1/1     Running     0             5m34s
kube-system                      kube-controller-manager-peer-pods-ctlplane-0     1/1     Running     1 (67s ago)   5m34s
kube-system                      kube-multus-ds-l7wcm                             1/1     Running     0             2m42s
kube-system                      kube-multus-ds-vgcl5                             1/1     Running     0             4m31s
kube-system                      kube-proxy-qpnct                                 1/1     Running     0             2m42s
kube-system                      kube-proxy-x6fzh                                 1/1     Running     0             5m21s
kube-system                      kube-scheduler-peer-pods-ctlplane-0              1/1     Running     1 (64s ago)   5m34s
root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
No resources found
root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
NAME        HANDLER     AGE
kata        kata        1s
kata-clh    kata-clh    2s
kata-qemu   kata-qemu   1s
```

Do you think it is reasonable to add this fix to this PR?

**Comment by snir911** [comment]
+1 for switching the PeerPods struct naming (not necessarily in this PR), PeerPodsInfra sounds suitable also maybe something like CoCoSystem similarly to the namespace could work.

**Comment by wainersm** [comment]
> The code looks reasonable to me, but I've tried testing it on three different VMs and each time hit:
> 
> ```
> Deploy Peer Pods
> Install the controller manager
> Wait for the cc-operator-controller-manager deployment be available
> Install CoCo and cloud-api-adaptor
> Wait for the pod cc-operator-daemon-install-lv4sd be ready
> Wait for the pod cloud-api-adaptor-daemonset-libvirt-ns2cv be ready
> Check the kata runtimeclass exists
> F0210 14:23:36.459155   78242 env.go:360] Setup failure: runtimeclasses.node.k8s.io "kata" not found
> ```
> 
> I think the issue is the runtime class is created slightly after the pods go into ready state, so we might need to add a wait for them to be created too?
> 

Hey @stevenhorsman , thanks for relentless testing this framework. You have found various important bugs!

I never hit that scenario (runtimeclass delayed to show up) on my environments. I will add a wait condition for the runtimeclass too.

> e.g. here is a run showing this:
> 
> ```
> FAIL
> make: *** [Makefile:66: test-e2e] Error 1
> root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
> No resources found
> root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get pods -A
> NAMESPACE                        NAME                                             READY   STATUS      RESTARTS      AGE
> autorules                        autoruler-76cdf5f9f6-d2pfx                       1/1     Running     0             4m31s
> confidential-containers-system   cc-operator-controller-manager-fbb5dcf9d-h27f7   2/2     Running     1 (67s ago)   112s
> confidential-containers-system   cc-operator-daemon-install-mgkcb                 1/1     Running     0             97s
> confidential-containers-system   cloud-api-adaptor-daemonset-libvirt-cr4nz        1/1     Running     0             97s
> ingress-nginx                    ingress-nginx-admission-create-qgrt8             0/1     Completed   0             4m32s
> ingress-nginx                    ingress-nginx-admission-patch-5rtg6              0/1     Completed   1             4m32s
> ingress-nginx                    ingress-nginx-controller-64f79ddbcc-v8nj6        1/1     Running     0             4m32s
> kube-flannel                     kube-flannel-ds-fwhmc                            1/1     Running     0             5m21s
> kube-flannel                     kube-flannel-ds-n7xt7                            1/1     Running     0             2m42s
> kube-system                      coredns-787d4945fb-882tr                         1/1     Running     0             5m21s
> kube-system                      coredns-787d4945fb-lgrjq                         1/1     Running     0             5m21s
> kube-system                      etcd-peer-pods-ctlplane-0                        1/1     Running     0             5m35s
> kube-system                      kube-apiserver-peer-pods-ctlplane-0              1/1     Running     0             5m34s
> kube-system                      kube-controller-manager-peer-pods-ctlplane-0     1/1     Running     1 (67s ago)   5m34s
> kube-system                      kube-multus-ds-l7wcm                             1/1     Running     0             2m42s
> kube-system                      kube-multus-ds-vgcl5                             1/1     Running     0             4m31s
> kube-system                      kube-proxy-qpnct                                 1/1     Running     0             2m42s
> kube-system                      kube-proxy-x6fzh                                 1/1     Running     0             5m21s
> kube-system                      kube-scheduler-peer-pods-ctlplane-0              1/1     Running     1 (64s ago)   5m34s
> root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
> No resources found
> root@stevenh-libvirt-2e2:~/go/src/github.com/wainersm/cc-cloud-api-adaptor# kubectl get runtimeclass
> NAME        HANDLER     AGE
> kata        kata        1s
> kata-clh    kata-clh    2s
> kata-qemu   kata-qemu   1s
> ```
>

Interesting. I was afraid that in the end the readyness checks for cc-operator-daemon-install and  cloud-api-adaptor-daemonset-libvirt were broken again (I had hard time figuring out the right conditions), which is not the case. The runtimeclass really show up a few seconds later.
 
> Do you think it is reasonable to add this fix to this PR?

Yes, I do.

File: test/e2e/README.md (hunk 1 above)
Comment (wainersm) [reply test/e2e/README.md:32]: Replaced the example with the @stevenhorsman 's.

**Comment by wainersm** [comment]
> > > @wainersm the name `PeerPods` for the struct handling cloud-api-adaptor (CAA) properties is confusing and overload. The PeerPod name refers to the pod. Once we have CRD to handle dangling resources the CRD's naming also most likely be PeerPod to reflect the pod. But in e2e PeerPod refers to CAA.
> > > I think renaming the PeerPod struct to something like PeerPodInfra will help.
> > 
> > 
> > Yes, I definitely understand where you are coming from here. `PeerPodInfra` is fine, but I wonder if `CloudApiAdaptor` would be as good/better
> 
> Yeah CloudAPIAdaptor is also fine and leaves no confusion :-). Some more thoughts below.
> 

Thanks for bringing this topic up @bpradipt. Naming is a art :)

I renamed the struct to `CloudAPIAdaptor`. Based on what you and @stevenhorsman mentioned, maybe that struct would be further broken into other concepts like `Coco` (to hold all CoCo-related properties and operations), however I think at this point we don't need to over-engineer it.

**Comment by wainersm** [comment]
Update this PR so that:
 - PeerPods is renamed to CloudAPIAdaptor (@bpradipt)
 - CloudAPIAdaptor.Deploy() wait the runtimeclass to show up (@stevenhorsman)
 - Clarified the example regarding the new TEST_E2E_TEARDOWN variable (@bpradipt , @stevenhorsman )

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @wainersm

**Comment by bpradipt** [comment]
@wainersm please update the README with the suggestion from @stevenhorsman and we can merge this one.

**Comment by stevenhorsman** [comment]
> @wainersm please update the README with the suggestion from @stevenhorsman and we can merge this one.

Hey @bpradipt - it seems like Wainer might not be available to make the change, so do you want me to push that 1 char update to the branch and we can get it merged rather than hold things up further for such a minor issue?

File: test/e2e/README.md [hunk 2]
```diff
@@ -26,6 +26,10 @@ The `TEST_E2E_PODVM_IMAGE` is an optional variable which specifies the path to t
 $ TEST_E2E_PROVISION=yes TEST_E2E_PODVM_IMAGE="path/to/podvm-base.qcow2" CLOUD_PROVIDER=libvirt make test-e2e
 ```
 
+By default it is given 20 minutes for the entire e2e execution to complete, otherwise the process is preempted. If you need to extend that timeout then export the `TEST_E2E_TIMEOUT` variable. For example, `TEST_E2E_TIMEOUT=30m` set the timeout to 30 minutes. See `-timeout` in [go test flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) for the values accepted.
+
+To leave the cluster untouched by the execution finish you should export `TEST_E2E_TEARDOWN=no`, otherwise the framework will attempt to wipe out any resources created. For example, if `TEST_E2E_PROVISION=yes` it used to create a VPC and cluster for testing and `TEST_E2E_TEARDOWN=no` not specified, then at the end of the test the provisioned cluster and VPC, will both be deleted.
```
[the remark this answers is not in the document]
Comment (wainersm) [reply test/e2e/README.md]: Done

**Comment by wainersm** [comment]
@bpradipt @stevenhorsman sorry folks, last weeks I got a few days off. I pushed the update to the README to fix the typo Steve found.

**Comment by bpradipt** [comment]
Looks like we are all set for this PR. Merging this.

**Pull request merged**